lib/plug/conn.ex

```diff
@@ -1328,50 +1340,149 @@ defmodule Plug.Conn do
+
+    cookies =
+      verify_or_decrypt(
+        opts[:signed],
+        req_cookies,
+        cookies,
+        &Plug.Crypto.verify(secret_key_base, &1 <> "_cookie", &2, keys: Plug.Keys)
+      )
+
+    cookies =
+      verify_or_decrypt(
+        opts[:encrypted],
+        req_cookies,
+        cookies,
+        &Plug.Crypto.decrypt(secret_key_base, &1 <> "_cookie", &2, keys: Plug.Keys)
+      )
+
+    %{conn | cookies: cookies}
+  end
+
+  defp verify_or_decrypt(names, req_cookies, cookies, fun) do
+    names
+    |> List.wrap()
+    |> Enum.reduce(cookies, fn name, acc ->
+      if value = req_cookies[name] do
+        case fun.(name, value) do
+          {:ok, verified_value} -> Map.put(acc, name, verified_value)
+          {_, _} -> Map.delete(acc, name)
+        end
+      else
+        acc
+      end
+    end)
+  end
+
   @doc """
   Puts a response cookie in the connection.
 
-  The cookie value is not automatically escaped. Therefore, if you
-  want to store values with comma, quotes, and so on, you need to explicitly
-  escape them or use a function such as `Base.encode64(value, padding: false)`
-  when writing and `Base.decode64(encoded, padding: false)` when reading
-  the cookie. Padding needs to be disabled since `=` is not a valid character
-  in cookie values.
+  The cookie value must be a binary and that the cookie value is not
+  automatically escaped, unless signing or encryption is enabled.
+  Therefore if you want to store values with non-alphanumeric characters,
+  you must either sign or encrypt the cookie (see the upcoming section)
+  or consider explicitly escaping the cookie value by using a function
+  such as `Base.encode64(value, padding: false)` when writing and
+  `Base.decode64(encoded, padding: false)` when reading the cookie.
+  It is important for padding to be disabled since `=` is not a valid
+  character in cookie values.
+
+  ## Signing and encrypting cookies
+
+  This function allows you to automatically sign and encrypt cookies.
+  When signing or encryption is enabled, then any Elixir value can be
+  stored in the cookie (except anonymous functions for security reasons).
+  Once a value is signed or encrypted, you must also call `fetch_cookie/2`
+  with the name of the cookies that are either signed or encrypted.
+
+  To sign, you would do:
+
+      put_resp_cookie(conn, "my-cookie", %{user_id: user.id}, sign: true)
+
+  and then:
+
+      fetch_cookie(conn, signed: ~w(my-cookie))
+
+  To encrypt, you would do:
+
+      put_resp_cookie(conn, "my-cookie", %{user_id: user.id}, encrypt: true)
+
+  and then:
+
+      fetch_cookie(conn, encrypted: ~w(my-cookie))
+
+  By default a signed or encrypted cookie is only valid for a day, unless
+  a `:max_age` is specified. The signing salt and encryption secret will
+  be the cookie name followed by the `_cookie` prefix. Be careful to not
+  reuse those as salts and secrets in other parts of the application.
```

Comment:
I would reword this a bit:
The key is derived from the connection's `secret_key_base` using a salt that is built by appending "_cookie" to the cookie name. Care should be taken not to derive other keys using this value as the salt.

Reply:
Thanks, done!